File: src/InfoResults.pas
unit InfoResults;

interface

uses Global, interface_global, interface_inforesults;

PROCEDURE WriteAssessmentSimulation(StrNr,totalnameEvalStat : string;
                                    TheProjectType : repTypeProject;
                                    RangeMin,RangeMax : LongInt);



implementation

Uses SysUtils;





PROCEDURE WriteAssessmentSimulation(StrNr,totalnameEvalStat : string;
                                    TheProjectType : repTypeProject;
                                    RangeMin,RangeMax : LongInt);
VAR fAssm : text;
    TypeObsSim : rep_TypeObsSim;
    Nobs,Nri : INTEGER;
    ObsAver,SimAver,PearsonCoeff,RMSE,NRMSE,NScoeff,IndexAg : double;
    ArrayObsSim : rep_ArrayObsSim;
    YearString : string;

BEGIN
// 1. Open file for assessment
Assign(fAssm,totalnameEvalStat);
Rewrite(fAssm);
WRITELN(fAssm,'AquaCrop 7.0 (June 2021) - Output created on (date) : ',DateToStr(Date),'   at (time) : ',TimeToStr(Time));
WRITELN(fAssm,'Evaluation of simulation results - Statistics');
IF (TheProjectType = TypePRM) THEN WRITELN(fAssm,'** Run number:',StrNr);
WRITELN(fAssm);

// 2. Run analysis

// 2.1 Canopy Cover
TypeObsSim := ObsSimCC;
StatisticAnalysis(TypeObsSim,RangeMin,RangeMax,StrNr,Nobs,ObsAver,SimAver,
                  PearsonCoeff,RMSE,NRMSE,NScoeff,IndexAg,ArrayObsSim);
                  WRITELN(fAssm);
WRITELN(fAssm,'  ASSESSMENT OF CANOPY COVER --------------------------------------');
IF (Nobs > 1)
   THEN BEGIN
        WRITELN(fAssm,'              --------- Canopy Cover (%) ---------');
        WRITELN(fAssm,'    Nr        Observed    +/- St Dev     Simulated    Date');
        WRITELN(fAssm,'  ----------------------------------------------------------------');
        FOR Nri := 1 to Nobs DO
            BEGIN
            IF (ArrayObsSim[Nri].YYYYi <= 1901)
               THEN YearString :=''
               ELSE Str((ArrayObsSim[Nri].YYYYi):4,YearString);
            WRITELN(fAssm,Nri:6,(ArrayObsSim[Nri].Obsi):14:1,(ArrayObsSim[Nri].StdObsi):14:1,
                         (ArrayObsSim[Nri].Simi):14:1,'      ',(ArrayObsSim[Nri].DDi):2,' ',
                         NameMonth[ArrayObsSim[Nri].MMi],' ',YearString);
            END;
        WRITELN(fAssm);
        WRITELN(fAssm,'  Valid observations/simulations sets (n) ....... : ',Nobs:5);
        WRITELN(fAssm,'  Average of observed Canopy Cover .............. : ',ObsAver:7:1,'   %');
        WRITELN(fAssm,'  Average of simulated Canopy Cover ............. : ',SimAver:7:1,'   %');
        WRITELN(fAssm);
        WRITELN(fAssm,'  Pearson Correlation Coefficient (r) ........... : ',PearsonCoeff:8:2);
        WRITELN(fAssm,'  Root mean square error (RMSE) ................. : ',RMSE:7:1,'   % CC');
        WRITELN(fAssm,'  Normalized root mean square error  CV(RMSE).... : ',NRMSE:7:1,'   %');
        WRITELN(fAssm,'  Nash-Sutcliffe model efficiency coefficient (EF): ',NScoeff:8:2);
        WRITELN(fAssm,'  Willmotts index of agreement (d) .............. : ',IndexAg:8:2);
        END
   ELSE WRITELN(fAssm,'  No statistic analysis (insufficient data)');
WRITELN(fAssm,'  ----------------------------------------------------------------');

//2.2 Biomass production
TypeObsSim := ObsSimB;
StatisticAnalysis(TypeObsSim,RangeMin,RangeMax,StrNr,Nobs,ObsAver,SimAver,
                  PearsonCoeff,RMSE,NRMSE,NScoeff,IndexAg,ArrayObsSim);
WRITELN(fAssm);
WRITELN(fAssm);
WRITELN(fAssm,'  ASSESSMENT OF BIOMASS PRODUCTION --------------------------------');
IF (Nobs > 1)
   THEN BEGIN
        WRITELN(fAssm,'              --------- Biomass (ton/ha) ---------');
        WRITELN(fAssm,'    Nr        Observed    +/- St Dev     Simulated    Date');
        WRITELN(fAssm,'  ----------------------------------------------------------------');
        FOR Nri := 1 to Nobs DO
            BEGIN
            IF (ArrayObsSim[Nri].YYYYi <= 1901)
               THEN YearString :=''
               ELSE Str((ArrayObsSim[Nri].YYYYi):4,YearString);
            WRITELN(fAssm,Nri:6,(ArrayObsSim[Nri].Obsi):16:3,(ArrayObsSim[Nri].StdObsi):14:3,
                          (ArrayObsSim[Nri].Simi):14:3,'      ',(ArrayObsSim[Nri].DDi):2,' ',
                          NameMonth[ArrayObsSim[Nri].MMi],' ',YearString);
            END;
        WRITELN(fAssm);
        WRITELN(fAssm,'  Valid observations/simulations sets (n) ....... : ',Nobs:5);
        WRITELN(fAssm,'  Average of observed Biomass production ........ : ',ObsAver:9:3,'   ton/ha');
        WRITELN(fAssm,'  Average of simulated Biomass production ....... : ',SimAver:9:3,'   ton/ha');
        WRITELN(fAssm);
        WRITELN(fAssm,'  Pearson Correlation Coefficient (r) ........... : ',PearsonCoeff:8:2);
        WRITELN(fAssm,'  Root mean square error (RMSE) ................. : ',RMSE:9:3,'   ton/ha');
        WRITELN(fAssm,'  Normalized root mean square error  CV(RMSE).... : ',NRMSE:7:1,'   %');
        WRITELN(fAssm,'  Nash-Sutcliffe model efficiency coefficient (EF): ',NScoeff:8:2);
        WRITELN(fAssm,'  Willmotts index of agreement (d) .............. : ',IndexAg:8:2);
        END
   ELSE WRITELN(fAssm,'  No statistic analysis (insufficient data)');
WRITELN(fAssm,'  ----------------------------------------------------------------');

//2.3 Soil Water Content
TypeObsSim := ObsSimSWC;
StatisticAnalysis(TypeObsSim,RangeMin,RangeMax,StrNr,Nobs,ObsAver,SimAver,
                  PearsonCoeff,RMSE,NRMSE,NScoeff,IndexAg,ArrayObsSim);
WRITELN(fAssm);
WRITELN(fAssm);
WRITELN(fAssm,'  ASSESSMENT OF SOIL WATER CONTENT --------------------------------');
IF (Nobs > 1)
   THEN BEGIN
        WRITELN(fAssm,'              ------ Soil water content (mm) -----');
        WRITELN(fAssm,'    Nr        Observed    +/- St Dev     Simulated    Date');
        WRITELN(fAssm,'  ----------------------------------------------------------------');
        FOR Nri := 1 to Nobs DO
            BEGIN
            IF (ArrayObsSim[Nri].YYYYi <= 1901)
               THEN YearString :=''
               ELSE Str((ArrayObsSim[Nri].YYYYi):4,YearString);
            WRITELN(fAssm,Nri:6,(ArrayObsSim[Nri].Obsi):14:1,(ArrayObsSim[Nri].StdObsi):14:1,
                      (ArrayObsSim[Nri].Simi):14:1,'      ',(ArrayObsSim[Nri].DDi):2,' ',
                      NameMonth[ArrayObsSim[Nri].MMi],' ',YearString);
            END;
        WRITELN(fAssm);
        WRITELN(fAssm,'  Valid observations/simulations sets (n) ....... : ',Nobs:5);
        WRITELN(fAssm,'  Average of observed Soil water content ........ : ',ObsAver:7:1,'   mm');
        WRITELN(fAssm,'  Average of simulated Soil water content ....... : ',SimAver:7:1,'   mm');
        WRITELN(fAssm);
        WRITELN(fAssm,'  Pearson Correlation Coefficient (r) ........... : ',PearsonCoeff:8:2);
        WRITELN(fAssm,'  Root mean square error (RMSE) ................. : ',RMSE:7:1,'   mm');
        WRITELN(fAssm,'  Normalized root mean square error  CV(RMSE).... : ',NRMSE:7:1,'   %');
        WRITELN(fAssm,'  Nash-Sutcliffe model efficiency coefficient (EF): ',NScoeff:8:2);
        WRITELN(fAssm,'  Willmotts index of agreement (d) .............. : ',IndexAg:8:2);
        END
   ELSE WRITELN(fAssm,'  No statistic analysis (insufficient data)');
WRITELN(fAssm,'  ----------------------------------------------------------------');
WRITELN(fAssm);

// 3. Close file for assessment
Close(fAssm);
END; (* WriteAssessmentSimulation *)



end.

File: src/inforesults.f90
module ac_inforesults

use ac_kinds, only: dp, &
                    int8, &
                    int16, &
                    int32, &
                    intEnum
use iso_fortran_env, only: iostat_end
use ac_global, only:    typeObsSim_ObsSimCC, &
                        typeObsSim_ObsSimB, &
                        typeObsSim_ObsSimSWC, &
                        undef_int, &
                        GetPathNameSimul, &
                        GetSimulation_FromDayNr, &
                        roundc
                     
implicit none

type rep_EventObsSim 
    real(dp) :: Obsi
        !! Undocumented
    real(dp) :: StdObsi
        !! Undocumented
    real(dp) :: Simi
        !! Undocumented
    integer(int8) :: DDi
        !! Undocumented
    integer(int8) :: MMi
        !! Undocumented
    integer(int32) :: YYYYi
        !! Undocumented
end type rep_EventObsSim 


contains

subroutine StatisticAnalysis(TypeObsSim, RangeObsMin, RangeObsMax, StrNr, &
                             Nobs, ObsAver, SimAver, PearsonCoeff, RMSE, &
                             NRMSE, NScoeff, IndexAg, ArrayObsSim)
    integer(intEnum), intent(in) :: TypeObsSim
    integer(int32), intent(in) :: RangeObsMin
    integer(int32), intent(in) :: RangeObsMax
    character(len=*), intent(in) :: StrNr
    integer(int32), intent(inout) :: Nobs
    real(dp), intent(inout) :: ObsAver
    real(dp), intent(inout) :: SimAver
    real(dp), intent(inout) :: PearsonCoeff
    real(dp), intent(inout) :: RMSE
    real(dp), intent(inout) :: NRMSE
    real(dp), intent(inout) :: NScoeff
    real(dp), intent(inout) :: IndexAg
    type(rep_EventObsSim), dimension(100), intent(inout) :: ArrayObsSim

    integer(int32) :: Nri
    real(dp) :: dDeNom, R2tel, SumSqrDobs, SumSqrDsim, SumSqrDiv, DeNom

    ! get data
    call GetObsSim(RangeObsMin, RangeObsMax, StrNr, Nobs, ArrayObsSim, &
                   ObsAver, SimAver)
    ! statistical evaluation
    if (Nobs > 1) then
        R2tel = 0._dp
        SumSqrDobs = 0._dp
        SumSqrDsim = 0._dp
        SumSqrDiv = 0._dp
        dDeNom = 0._dp
        do Nri = 1, Nobs 
            R2tel = R2tel + (ArrayObsSim(Nri)%Obsi-ObsAver) &
                            * (ArrayObsSim(Nri)%Simi-SimAver)
            SumSqrDobs = SumSqrDobs + (ArrayObsSim(Nri)%Obsi-ObsAver)**2
            SumSqrDsim = SumSqrDsim + (ArrayObsSim(Nri)%Simi-SimAver)**2
            SumSqrDiv = SumSqrDiv &
                        + (ArrayObsSim(Nri)%Obsi-ArrayObsSim(Nri)%Simi)**2
            dDeNom = dDeNom + (abs(ArrayObsSim(Nri)%Simi-ObsAver) &
                               + abs(ArrayObsSim(Nri)%Obsi-ObsAver))**2
        end do
        ! R2
        DeNom = sqrt(SumSqrDobs*SumSqrDsim)
        if (DeNom > 0._dp) then
            PearsonCoeff = R2tel/DeNom
        else
            PearsonCoeff = real(undef_int, kind=dp)
        end if
        ! RMSE
        RMSE = sqrt(SumSqrDiv/NObs)
        ! NRMSE
        if (ObsAver > 0._dp) then
            NRMSE = 100._dp*(RMSE/ObsAver)
        else
            NRMSE = real(undef_int, kind=dp)
        end if
        ! Nash-Sutcliffe coefficient (EF)
        if (SumSqrDobs > 0._dp) then
            NScoeff = 1._dp - (SumSqrDiv/SumSqrDobs)
        else
            NScoeff = real(undef_int, kind=dp)
        end if
        ! Index of agreement (d)
        if (dDeNom > 0._dp) then
            IndexAg = 1._dp - (SumSqrDiv/dDeNom)
        else
            IndexAg = real(undef_int, kind=dp)
        end if
    else
        ObsAver = real(undef_int, kind=dp)
        SimAver = real(undef_int, kind=dp)
        PearsonCoeff = real(undef_int, kind=dp)
        RMSE = real(undef_int, kind=dp)
        NRMSE = real(undef_int, kind=dp)
        NScoeff = real(undef_int, kind=dp)
        IndexAg = real(undef_int, kind=dp)
    end if


    contains


    subroutine GetObsSim(RangeObsMin, RangeObsMax, StrNr, Nobs, &
                         ArrayObsSim, ObsAver, SimAver)
        integer(int32), intent(in) :: RangeObsMin
        integer(int32), intent(in) :: RangeObsMax
        character(len=*), intent(in) :: StrNr
        integer(int32), intent(inout) :: Nobs
        type(rep_EventObsSim), dimension(100), intent(inout) :: ArrayObsSim
        real(dp), intent(inout) :: ObsAver
        real(dp), intent(inout) :: SimAver

        character(len=:), allocatable :: OutputName
        integer :: fOut
        integer(int8) :: Dayi, Monthi
        integer(int32) :: SkipLines, NCi, NCobs, Yeari
        integer(int32) :: i, rc
        real(dp) :: VarObsi, VarSimi, VarStdi

        Nobs = 0
        ObsAver = 0._dp
        SimAver = 0._dp

        ! open file
        OutputName = GetPathNameSimul() // 'EvalData' // trim(StrNr) // '.OUT'
        open(newunit=fOut, file=trim(OutputName), status='old', action='read')
        read(fOut, *) ! AquaCrop Version - Date and Time
        read(fOut, *) ! title
        read(fOut, *) ! 
        read(fOut, *) ! list of variables
        read(fOut, *) ! units
        ! find first day
        SkipLines = RangeObsMin - GetSimulation_FromDayNr()
        do i = 1, SkipLines 
            read(fOut, *, iostat = rc)
        end do

        ! get Sim and Obs in range
        select case (TypeObsSim)
            case(typeObsSim_ObsSimCC)
                NCobs = 2
            case (typeObsSim_ObsSimB)
                NCobs = 5
            case (typeObsSim_ObsSimSWC)
                NCobs = 8
            case default
                NCobs = 5
        end select
        do i = RangeObsMin, RangeObsMax 
            if (rc /= iostat_end) then
                read(fOut, *) Dayi, Monthi, Yeari
                do NCi = 1, NCobs 
                    read(fOut, *) VarSimi
                end do
                read(fOut, *) VarSimi, VarObsi, VarStdi
                if ((roundc(VarObsi, mold=1) /= undef_int) &
                                .and. (Nobs < 100)) then
                    Nobs = Nobs + 1
                    ArrayObsSim(Nobs)%DDi = Dayi
                    ArrayObsSim(Nobs)%MMi = Monthi
                    ArrayObsSim(Nobs)%YYYYi = Yeari
                    ArrayObsSim(Nobs)%Simi = VarSimi
                    ArrayObsSim(Nobs)%Obsi = VarObsi
                    ArrayObsSim(Nobs)%StdObsi = VarStdi
                    SimAver = SimAver + VarSimi
                    ObsAver = ObsAver + VarObsi
                end if
            end if
        end do
        ! close file
        close(fOut)

        ! calculate averages
        if (Nobs > 0) then
            ObsAver = ObsAver/real(Nobs, kind=dp)
            SimAver = SimAver/real(Nobs, kind=dp)
        end if
    end subroutine GetObsSim
end subroutine StatisticAnalysis



end module ac_inforesults
